nomos-services/membership/src/backends/mod.rs

```diff
@@ -0,0 +1,40 @@
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use nomos_sdp::backends::FinalizedBlockEvent;
+use nomos_sdp_core::{BlockNumber, DeclarationUpdate, ProviderInfo, ServiceType};
+use overwatch::DynError;
+
+pub mod sdp;
+
+pub struct MembershipBackendSettings {
+    pub settings_per_service: HashMap<ServiceType, SnapshotSettings>,
+}
+
+pub enum SnapshotSettings {
+    Block(usize),
+}
+
+#[derive(Debug, Clone)]
+pub struct MembershipEntry {
+    pub block_number: BlockNumber,
+    pub data: HashMap<ServiceType, HashMap<ProviderInfo, DeclarationUpdate>>,
+}
+
+#[async_trait]
+pub trait MembershipBackend {
+    // todo: introduce error type instead of DynError
+    fn init(settings: MembershipBackendSettings) -> Self;
+
+    // Get the snapshot at the given index for the specified service type.
+    // The snapshot index a sequential order of the service type specific time
+    // snapshots. index = 0 is the latest snapshot
+    // index > 0 is the index of the snapshot in the past
+    // index < 0 is the index of the snapshot relative to the latest snapshot
+    async fn get_snapshot_at(
+        &self,
+        service_type: ServiceType,
+        index: i32,
+    ) -> Result<HashMap<ProviderInfo, DeclarationUpdate>, DynError>;
```

Comment:
We could have `MembershipError` alongside `MembershiptEntry`, it would be more convenient to act on errors on receiving end.

Reply:
I did not implement exact type and error yet, this was still in progress so I can add what I encounter and modify, during implementation of the backend. I will define the error type as well, we will need it for sure, and modify variants in different PRs if needed.